service/tablet_allocator.cc

```diff
@@ -305,8 +411,79 @@ class load_balancer {
             lblogger.info("Prepared {} migrations in DC {}", dc_plan.size(), dc);
             plan.merge(std::move(dc_plan));
         }
+        plan.merge(co_await make_resize_plan());
 
-        lblogger.info("Prepared {} migrations", plan.size());
+        lblogger.info("Prepared {} migration plans, out of which, tablet_migrations={} and resize_requests={}",
+                      plan.size(), plan.tablet_migrations(), plan.resize_requests());
+        co_return std::move(plan);
+    }
+
+    future<migration_plan> make_resize_plan() {
+        migration_plan plan;
+
+        if (!_tm->tablets().balancing_enabled()) {
+            co_return std::move(plan);
+        }
+
+        cluster_resize_stats resize_load;
+
+        for (auto&& [table, tmap_] : _tm->tablets().all_tables()) {
+            auto& tmap = tmap_;
+
+            auto avg_tablet_size = _stats.load_stats_for_table(table).size_in_bytes / std::max(tmap.tablet_count(), size_t(1));
+            resize_load.update(table, avg_tablet_size, tmap.resize());
+            lblogger.info("Table {} with tablet_count={} has an average tablet size of {}", table, tmap.tablet_count(), avg_tablet_size);
+            co_await coroutine::maybe_yield();
+        }
+
+        // Emit new resize decisions
+
+        table_resize_plan resize_plan;
+
+        auto resize_units = ssize_t(max_concurrent_resize_requests) - ssize_t(resize_load.tables_being_resized.size());
```

Comment:
Isn't max_concurrent_resize_requests == 1, the resize_units could be negative.

Reply:
that's deliberate. if units is negative, then no resize is allowed. shouldn't happen in practice though.